server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -210,44 +189,30 @@ protected boolean defaultDocValues(Version indexCreated) {
             return fieldType.tokenized() == false;
         }
 
-        protected void setupFieldType(BuilderContext context) {
-            fieldType.setName(buildFullName(context));
-            if (fieldType.indexAnalyzer() == null && fieldType.tokenized() == false && fieldType.indexOptions() != IndexOptions.NONE) {
-                fieldType.setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
-                fieldType.setSearchAnalyzer(Lucene.KEYWORD_ANALYZER);
-            }
-            boolean defaultDocValues = defaultDocValues(context.indexCreatedVersion());
-            defaultFieldType.setHasDocValues(defaultDocValues);
-            if (docValuesSet == false) {
-                fieldType.setHasDocValues(defaultDocValues);
-            }
-        }
-
         /** Set metadata on this field. */
         public T meta(Map<String, String> meta) {
-            fieldType.setMeta(meta);
+            this.meta = meta;
             return (T) this;
         }
     }
 
     protected final Version indexCreatedVersion;
-    protected MappedFieldType fieldType;
-    protected final MappedFieldType defaultFieldType;
+    protected FieldType fieldType;
+    protected MappedFieldType mappedFieldType;
     protected MultiFields multiFields;
     protected CopyTo copyTo;
 
-    protected FieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+    protected FieldMapper(String simpleName, FieldType fieldType, MappedFieldType mappedFieldType,
```

Comment:
How hard would it be to get rid of `FieldType` here? It's a bit annoying because usually there isn't a single `FieldType` that applies, e.g. numeric fields leverage both points and numeric doc values, but it's impossible to configure both on a `FieldType`, which is why we create different `Field`s for points and doc values.

Reply:
Oh apologies, I just saw your previous message where you explain exactly this. +1 for progress over perfection and keeping it for a follow-up refactoring.